gpgme fails to build on macOS

**Describe the bug**
gpgme fails to build on macOS

**To Reproduce**

<details>
<summary><code>nix-build . -A gpgme</code> (<i>expand to see the output</i>)</summary>

<pre>
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
this derivation will be built:
  /nix/store/zwf2jgy24qs7b69n1rb5pg9ywwmi82f0-gpgme-1.15.0.drv
building '/nix/store/zwf2jgy24qs7b69n1rb5pg9ywwmi82f0-gpgme-1.15.0.drv'...
unpacking sources
unpacking source archive /nix/store/sq6ypkha64hcskq8qqw8rv40d9k4y94f-gpgme-1.15.0.tar.bz2
source root is gpgme-1.15.0
setting SOURCE_DATE_EPOCH to timestamp 1605173405 of file gpgme-1.15.0/gpgme.spec
patching sources
applying patch /nix/store/d5svj9vzmvfvlfh67chy0yliwjk50wb5-c4cf527ea227edb468a84bf9b8ce996807bd6992.patch
patching file src/engine-gpg.c
Hunk #1 succeeded at 2989 (offset 28 lines).
applying patch /nix/store/v28z94l2z3669df0vzwgsi4y53rvv5rc-0001-Fix-python-tests-on-non-Linux.patch
patching file configure.ac
patching file lang/python/tests/Makefile.am
applying patch /nix/store/k3dx8r5bp3pqd4dzrydj4yywg6c9p6ja-disable-python-tests.patch
patching file lang/python/Makefile.am
autoreconfPhase
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: copying file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: aclocal --force -I m4
autoreconf: running: /nix/store/ymq2bmizcmzffq8qwi11lpm2sfl0d9r2-autoconf-2.70/bin/autoconf --force
configure.ac:100: warning: The macro `AC_CONFIG_HEADER' is obsolete.
configure.ac:100: You should run autoupdate.
./lib/autoconf/status.m4:719: AC_CONFIG_HEADER is expanded from...
configure.ac:100: the top level
configure.ac:108: warning: The macro `AC_GNU_SOURCE' is obsolete.
configure.ac:108: You should run autoupdate.
./lib/autoconf/specific.m4:312: AC_GNU_SOURCE is expanded from...
configure.ac:108: the top level
configure.ac:246: warning: The macro `AC_TRY_RUN' is obsolete.
configure.ac:246: You should run autoupdate.
./lib/autoconf/general.m4:2996: AC_TRY_RUN is expanded from...
m4/glib-2.0.m4:8: AM_PATH_GLIB_2_0 is expanded from...
configure.ac:246: the top level
configure.ac:246: warning: The macro `AC_TRY_LINK' is obsolete.
configure.ac:246: You should run autoupdate.
./lib/autoconf/general.m4:2919: AC_TRY_LINK is expanded from...
m4/glib-2.0.m4:8: AM_PATH_GLIB_2_0 is expanded from...
configure.ac:246: the top level
configure.ac:248: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:248: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:248: the top level
configure.ac:296: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:296: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:296: the top level
configure.ac:309: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:309: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:309: the top level
configure.ac:499: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:499: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:499: the top level
configure.ac:518: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:518: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:518: the top level
configure.ac:524: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:524: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:524: the top level
configure.ac:530: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:530: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:530: the top level
configure.ac:536: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:536: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:536: the top level
configure.ac:705: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:705: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
m4/gpg-error.m4:27: AM_PATH_GPG_ERROR is expanded from...
configure.ac:705: the top level
configure.ac:712: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:712: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
m4/libassuan.m4:17: _AM_PATH_LIBASSUAN_COMMON is expanded from...
m4/libassuan.m4:158: AM_PATH_LIBASSUAN is expanded from...
configure.ac:712: the top level
configure.ac:755: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:755: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:755: the top level
configure.ac:787: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:787: You should run autoupdate.
./lib/autoconf/general.m4:203: AC_HELP_STRING is expanded from...
configure.ac:787: the top level
autoreconf: running: /nix/store/ymq2bmizcmzffq8qwi11lpm2sfl0d9r2-autoconf-2.70/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:108: installing 'build-aux/compile'
configure.ac:101: installing 'build-aux/missing'
lang/cpp/src/Makefile.am: installing 'build-aux/depcomp'
autoreconf: 'build-aux/config.sub' is updated
autoreconf: 'build-aux/config.guess' is updated
autoreconf: Leaving directory '.'
configuring
fixing libtool script ./build-aux/ltmain.sh
configure flags: --disable-static --disable-dependency-tracking --prefix=/nix/store/9357hqrwqm6wnvl6qnp1hw2h045nw2rf-gpgme-1.15.0 --bindir=/nix/store/r9355cgd5y62qh03h68b4a6lj3w3kcap-gpgme-1.15.0-dev/bin --sbindir=/nix/store/r9355cgd5y62qh03h68b4a6lj3w3kcap-gpgme-1.15.0-dev/sbin --includedir=/nix/store/r9355cgd5y62qh03h68b4a6lj3w3kcap-gpgme-1.15.0-dev/include --oldincludedir=/nix/store/r9355cgd5y62qh03h68b4a6lj3w3kcap-gpgme-1.15.0-dev/include --mandir=/nix/store/r9355cgd5y62qh03h68b4a6lj3w3kcap-gpgme-1.15.0-dev/share/man --infodir=/nix/store/a7im3pbpc9aa0v37wlclqnas2j5cx5kv-gpgme-1.15.0-info/share/info --docdir=/nix/store/9357hqrwqm6wnvl6qnp1hw2h045nw2rf-gpgme-1.15.0/share/doc/gpgme --libdir=/nix/store/9357hqrwqm6wnvl6qnp1hw2h045nw2rf-gpgme-1.15.0/lib --libexecdir=/nix/store/9357hqrwqm6wnvl6qnp1hw2h045nw2rf-gpgme-1.15.0/libexec --localedir=/nix/store/9357hqrwqm6wnvl6qnp1hw2h045nw2rf-gpgme-1.15.0/share/locale --enable-fixed-path=/nix/store/x5v8m4xf9sf07675db7cfi19g9qjp4y1-gnupg-2.2.24/bin --with-libgpg-error-prefix=/nix/store/4sdr1n9fvhpry7410jfyd87sdsvh5422-libgpg-error-1.38-dev --with-libassuan-prefix=/nix/store/c0pfl1d4dw61i3mwzc52akb9vbx223vf-libassuan-2.5.4-dev --disable-gpg-test
checking for a BSD-compatible install... /nix/store/5m4875pw1zcd588n7a6j88vwsilvnm7j-coreutils-8.32/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /nix/store/5m4875pw1zcd588n7a6j88vwsilvnm7j-coreutils-8.32/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking build system type... x86_64-apple-darwin20.2.0
checking host system type... x86_64-apple-darwin20.2.0
checking whether make supports nested variables... (cached) yes
checking whether make supports the include directive... yes (GNU style)
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether clang accepts -g... yes
checking for clang option to enable C11 features... none needed
checking whether clang understands -c and -o together... yes
checking dependency style of clang... none
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking for locale.h... yes
checking for sys/select.h... yes
checking for sys/uio.h... yes
checking for argp.h... no
checking for sys/time.h... yes
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for gcc... (cached) clang
checking whether the compiler supports GNU C... (cached) yes
checking whether clang accepts -g... (cached) yes
checking for clang option to enable C11 features... (cached) none needed
checking whether clang understands -c and -o together... (cached) yes
checking dependency style of clang... (cached) none
checking how to run the C preprocessor... clang -E
checking whether the compiler supports GNU C++... yes
checking whether clang++ accepts -g... yes
checking for clang++ option to enable C++11 features... none needed
checking dependency style of clang++... none
checking for gitlog-to-changelog... no
checking for cc for build... clang
checking how to print strings... printf
checking for a sed that does not truncate output... /nix/store/n0zyx5b5s1grq0jrcy2xvf7apvh25wry-gnused-4.8/bin/sed
checking for grep that handles long lines and -e... /nix/store/pmbv3czhsa816msq51ngb2db61crjqw0-gnugrep-3.6/bin/grep
checking for egrep... /nix/store/pmbv3czhsa816msq51ngb2db61crjqw0-gnugrep-3.6/bin/grep -E
checking for fgrep... /nix/store/pmbv3czhsa816msq51ngb2db61crjqw0-gnugrep-3.6/bin/grep -F
checking for ld used by clang... ld
checking if the linker (ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... nm
checking the name lister (nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 786432
checking how to convert x86_64-apple-darwin20.2.0 file names to x86_64-apple-darwin20.2.0 format... func_convert_file_noop
checking how to convert x86_64-apple-darwin20.2.0 file names to toolchain format... func_convert_file_noop
checking for ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... no
checking for strip... strip
checking for ranlib... ranlib
checking command to parse nm output from clang object... ok
checking for sysroot... no
checking for a working dd... /nix/store/5m4875pw1zcd588n7a6j88vwsilvnm7j-coreutils-8.32/bin/dd
checking how to truncate binary pipes... /nix/store/5m4875pw1zcd588n7a6j88vwsilvnm7j-coreutils-8.32/bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking for dsymutil... dsymutil
checking for nmedit... no
checking for lipo... lipo
checking for otool... otool
checking for otool64... no
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking for -force_load linker flag... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if clang supports -fno-rtti -fno-exceptions... yes
checking for clang option to produce PIC... -fno-common -DPIC
checking if clang PIC flag -fno-common -DPIC works... yes
checking if clang static flag -static works... no
checking if clang supports -c -o file.o... yes
checking if clang supports -c -o file.o... (cached) yes
checking whether the clang linker (ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin20.2.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
checking how to run the C++ preprocessor... clang++ -E
checking for ld used by clang++... ld
checking if the linker (ld) is GNU ld... no
checking whether the clang++ linker (ld) supports shared libraries... yes
checking for clang++ option to produce PIC... -fno-common -DPIC
checking if clang++ PIC flag -fno-common -DPIC works... yes
checking if clang++ static flag -static works... no
checking if clang++ supports -c -o file.o... yes
checking if clang++ supports -c -o file.o... (cached) yes
checking whether the clang++ linker (ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin20.2.0 dyld
checking how to hardcode library paths into programs... immediate
checking for windres... no
checking whether clang++ supports C++11 features by default... yes
checking pkg-config is at least version 0.9.0... yes
checking for GPGME_QT... no
checking for GPGME_QTTEST... no
Package Qt5Core was not found in the pkg-config search path.
Perhaps you should add the directory containing `Qt5Core.pc'
to the PKG_CONFIG_PATH environment variable
No package 'Qt5Core' found
configure: WARNING:
***
*** Qt5 (Qt5Core) not found Qt Binding will be disabled.
***
checking for swig... no
checking for swig2.0... no
checking for inline... inline
checking size of unsigned int... 4
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for off_t... yes
checking for uintptr_t... yes
checking if gcc supports -Wno-missing-field-initializers... yes
checking for __thread... yes
configure: checking for libraries
checking for _LARGEFILE_SOURCE value needed for large files... no
checking how clang reports undeclared, standard C functions... configure: error: in `/private/tmp/nix-build-gpgme-1.15.0.drv-0/gpgme-1.15.0':
configure: error: cannot detect from compiler exit status or warnings
See `config.log' for more details
error: --- Error --- nix-build
builder for '/nix/store/zwf2jgy24qs7b69n1rb5pg9ywwmi82f0-gpgme-1.15.0.drv' failed with exit code 1
</pre>
</details>

**Expected behavior**

To build successfully.

**Additional context**

I'm working on my flake based config with nix, home-manager & nix-darwin. I updated my inputs today `nix flake update --recreate-lock-file` & after that, I couldn't build anymore. I cloned this repo & ran `nix-build . -A gpgme` which failed with the same error from `nix build './#foo'` 

I also tried this workaround https://github.com/NixOS/nixpkgs/issues/60313#issuecomment-499921195 but it didn't help.

**Notify maintainers**
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@primeos

**Metadata**
```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.2.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.4pre20201201_5a6ddb3`
 - channels(ahmed): `"darwin, nixpkgs-21.03pre260775.2080afd0399"`
```

--- a/pkgs/development/libraries/gpgme/default.nix
+++ b/pkgs/development/libraries/gpgme/default.nix
@@ -34,7 +34,9 @@ stdenv.mkDerivation rec {
       sha256 = "00d4sxq63601lzdp2ha1i8fvybh7dzih4531jh8bx07fab3sw65g";
     })
     # Disable python tests on Darwin as they use gpg (see configureFlags below)
-  ] ++ lib.optional stdenv.isDarwin ./disable-python-tests.patch;
+  ] ++ lib.optional stdenv.isDarwin ./disable-python-tests.patch
+  # Fix _AC_UNDECLARED_WARNING for autoconf≥2.70. See https://lists.gnupg.org/pipermail/gnupg-devel/2020-November/034643.html
+  ++ lib.optional stdenv.cc.isClang ./fix-clang-autoconf-undeclared-warning.patch;
 
   outputs = [ "out" "dev" "info" ];
   outputBin = "dev"; # gpgme-config; not so sure about gpgme-tool

--- a/pkgs/development/libraries/gpgme/fix-clang-autoconf-undeclared-warning.patch
+++ b/pkgs/development/libraries/gpgme/fix-clang-autoconf-undeclared-warning.patch
@@ -1,0 +1,36 @@
+diff -Naur gpgme.old/configure.ac gpgme.new/configure.ac
+--- gpgme.old/configure.ac	2020-11-12 04:19:50.000000000 -0500
++++ gpgme.new/configure.ac	2021-01-08 03:04:38.000000000 -0500
+@@ -166,6 +166,16 @@
+                           mym4_minor mym4_micro)
+ AC_SUBST(VERSION_NUMBER)
+ 
++# Try to find a thread-safe version of ttyname().
++gnupg_REPLACE_TTYNAME_R
++if test "$ac_cv_func_ttyname_r" != yes; then
++  AC_MSG_WARN([
++***
++*** ttyname() is not thread-safe and ttyname_r() does not exist
++***])
++fi
++
++
+ # We need to compile and run a program on the build machine.  A
+ # comment in libgpg-error says that the AC_PROG_CC_FOR_BUILD macro in
+ # the AC archive is broken for autoconf 2.57.  Given that there is no
+@@ -658,15 +668,6 @@
+ 
+ AC_FUNC_FSEEKO
+ 
+-# Try to find a thread-safe version of ttyname().
+-gnupg_REPLACE_TTYNAME_R
+-if test "$ac_cv_func_ttyname_r" != yes; then
+-  AC_MSG_WARN([
+-***
+-*** ttyname() is not thread-safe and ttyname_r() does not exist
+-***])
+-fi
+-
+ # Try to find a thread-safe version of getenv().
+ have_thread_safe_getenv=no
+ jm_GLIBC21

